dev/node_func.go

```diff
@@ -33,12 +33,22 @@ func (n *Node) Port() string {
 }
 
 func (n *Node) String() string {
-	n.mu.Lock()
-	defer n.mu.Unlock()
-	return fmt.Sprintf(
-		"node %d | addr: %s | latency: %v",
-		n.id, n.addr, n.latency,
-	)
+	if n != nil {
+		return fmt.Sprintf("addr: %s", n.addr)
+	}
+	return ""
```

Comment:
Maybe `return "<nil>"` to be explicit about that the `n` is nil?
An empty string is easy to miss in the output.
Ditto for line 51.